File: cousinsBT.cpp
// Time Complexity : O(N)
// Space Complexity : O(N)
// Leetcode passes: yes
// Any problems while solving this problem : no

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        
        queue<TreeNode*> q;
        if(root == NULL)
            return false;
        
        q.push(root);
        while(!q.empty())
        {
            int size = q.size();
            bool is_x = false;
            bool is_y = false;
            for(int i = 0; i < size; i++)
            {
                TreeNode* curr = q.front();
                q.pop();
                if(curr->val == x) is_x = true;
                if(curr->val == y) is_y = true;
                
                if(curr->left and curr->right)
                {
                    if((curr->left->val == x and curr->right->val == y) or (curr->left->val == y and curr->right->val == x)) return false;
                }
                if(curr->left) q.push(curr->left);
                if(curr->right) q.push(curr->right);
            }
            if(is_x and is_y) return true;
        }
        return false;
        
    }
};

File: cousinsBT.py
# // Time Complexity : O(n)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode : Yes
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    dx = 0
    px = 0
    dy = 0
    py = 0
    def isCousins(self, root, x, y):
        """
        :type root: TreeNode
        :type x: int
        :type y: int
        :rtype: bool
        """
        def findDepth(root,x,y,depth,parent):
            if root:
                if root.val == x:
                    self.dx = depth
                    self.px = parent
                    # print(dx)
                if root.val == y:
                    self.dy = depth
                    self.py = parent
                    # print(dy)

                findDepth(root.left,x,y,depth+1,root.val)
                findDepth(root.right,x,y,depth+1,root.val)




        # print(d1,d2)
        findDepth(root,x,y,0,0)
        # print(dx,px,dy,py)
        if self.dx==self.dy and self.px!=self.py:
            return True
        else:
            return False










